site/server/app/com/fortysevendeg/exercises/services/UserServices.scala

```diff
@@ -1,71 +1,79 @@
-
-  def createUser(request: CreateUserRequest): Future[CreateUserResponse]
+import com.fortysevendeg.exercises.models.{ UserStore }
+import shared.User
 
+trait UserServices {
+  def all: List[User]
+
+  def getUserByLogin(login: String): Option[User]
+
+  def getUserOrCreate(
+    login:       String,
+    name:        String,
+    github_id:   String,
+    picture_url: String,
+    github_url:  String,
+    email:       String
+  ): Throwable Xor User
+
+  def createUser(
+    login:       String,
+    name:        String,
+    github_id:   String,
+    picture_url: String,
+    github_url:  String,
+    email:       String
+  ): Throwable Xor User
+
+  def update(user: User): Boolean
+
+  def delete(user: User): Boolean
 }
 
-class UserServiceImpl(userStore: UserStore)(implicit val executionContext: ExecutionContext) extends UserServices {
-
-  override def getUserOrCreate(request: GetUserOrCreateRequest): Future[GetUserOrCreateResponse] =
-
-    getUserByLogin(GetUserByLoginRequest(request.login))
-      .flatMap {
-        _.user
-          .map(u ⇒ Future.successful(u))
-          .getOrElse(
-            createUser(CreateUserRequest(
-              login = request.login,
-              name = request.name,
-              github_id = request.github_id,
-              picture_url = request.picture_url,
-              github_url = request.github_url,
-              email = request.email
-            ))
-              .map(_.user)
-          )
-          .map(GetUserOrCreateResponse)
-      }
-
-  override def getUserByLogin(request: GetUserByLoginRequest): Future[GetUserByLoginResponse] = {
-
-    val result = for {
-      user ← userStore.getByLogin(login = request.login)
-    } yield GetUserByLoginResponse(user = user.headOption)
-
-    result recover {
-      case e ⇒
-        throw new Exception(s"User fetching error: ${e.getMessage}")
+class UserServiceImpl(implicit userStore: UserStore) extends UserServices {
+  def all: List[User] =
+    userStore.all
+
+  def getUserByLogin(login: String): Option[User] =
+    userStore.getByLogin(login)
+
+  def getUserOrCreate(
+    login:       String,
+    name:        String,
+    github_id:   String,
+    picture_url: String,
+    github_url:  String,
+    email:       String
+  ): Throwable Xor User = {
+    getUserByLogin(login) match {
+      case Some(user) ⇒ Xor.Right(user)
+      case _          ⇒ createUser(login, name, github_id, picture_url, github_url, email)
     }
   }
 
-  override def createUser(request: CreateUserRequest): Future[CreateUserResponse] = {
+  def createUser(
+    login:       String,
+    name:        String,
+    github_id:   String,
+    picture_url: String,
+    github_url:  String,
+    email:       String
+  ): Throwable Xor User = {
+    val user = User(None, login, name, github_id, picture_url, github_url, email)
+    Xor.fromOption(userStore.create(user), new Exception("Couldn't create user"))
```

Comment:
I think we don't want generic exceptions. We could wrap this exception in a more specific one.